lightning/src/routing/scorer.rs

```diff
@@ -127,55 +179,106 @@ impl Default for ScoringParameters {
-			Some((penalty_msat, last_failure)) => self.decay_from(*penalty_msat, last_failure),
-			None => 0,
-		};
-		#[cfg(feature = "no-std")]
-		let failure_penalty_msat =
-			self.channel_failures.get(&short_channel_id).copied().unwrap_or(0);
+		let failure_penalty_msat = self.channel_failures
+			.get(&short_channel_id)
+			.map_or(0,|failure| failure.decayed_penalty(self.params.failure_penalty_half_life));
 
 		self.params.base_penalty_msat + failure_penalty_msat
 	}
 
 	fn payment_path_failed(&mut self, _path: &Vec<RouteHop>, short_channel_id: u64) {
 		let failure_penalty_msat = self.params.failure_penalty_msat;
-		#[cfg(not(feature = "no-std"))]
-		{
-			let half_life = self.params.failure_penalty_half_life;
-			self.channel_failures
-				.entry(short_channel_id)
-				.and_modify(|(penalty_msat, last_failure)| {
-					let decayed_penalty = decay_from(*penalty_msat, last_failure, half_life);
-					*penalty_msat = decayed_penalty + failure_penalty_msat;
-					*last_failure = Instant::now();
-				})
-				.or_insert_with(|| (failure_penalty_msat, Instant::now()));
-		}
-		#[cfg(feature = "no-std")]
+		let half_life = self.params.failure_penalty_half_life;
 		self.channel_failures
 			.entry(short_channel_id)
-			.and_modify(|penalty_msat| *penalty_msat += failure_penalty_msat)
-			.or_insert(failure_penalty_msat);
+			.and_modify(|failure| failure.add_penalty(failure_penalty_msat, half_life))
+			.or_insert_with(|| ChannelFailure::new(failure_penalty_msat));
 	}
 }
 
 #[cfg(not(feature = "no-std"))]
-fn decay_from(penalty_msat: u64, last_failure: &Instant, half_life: Duration) -> u64 {
-	let decays = last_failure.elapsed().as_secs().checked_div(half_life.as_secs());
-	match decays {
-		Some(decays) => penalty_msat >> decays,
-		None => 0,
+impl Clock for std::time::Instant {
+	fn now() -> Self {
+		std::time::Instant::now()
 	}
+
+	fn elapsed(&self) -> Duration {
+		std::time::Instant::elapsed(self)
+	}
+}
+
+/// A clock that is always present, now or after time has passed.
+pub struct AlwaysPresent;
+
+impl Clock for AlwaysPresent {
+	fn now() -> Self {
+		Self
+	}
+
+	fn elapsed(&self) -> Duration {
+		Duration::from_secs(0)
+	}
+}
+
+impl Sub<Duration> for AlwaysPresent {
+	type Output = Self;
+
+	fn sub(self, _other: Duration) -> Self {
+		self
+	}
+}
+
+impl<C: Clock> Writeable for ScorerUsingClock<C> {
+	#[inline]
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
+		self.params.write(w)?;
+		self.channel_failures.write(w)
+	}
+}
+
+impl<C: Clock> Readable for ScorerUsingClock<C> {
+	#[inline]
+	fn read<R: Read>(r: &mut R) -> Result<Self, DecodeError> {
+		Ok(Self {
+			params: Readable::read(r)?,
+			channel_failures: Readable::read(r)?,
+		})
+	}
+}
+
+impl<C: Clock> Writeable for ChannelFailure<C> {
+	#[inline]
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
+		self.undecayed_penalty_msat.write(w)?;
+		(duration_since_epoch() - self.last_failed.elapsed()).write(w)
```

Comment:
Hmm, I wonder if we should document somewhere that an LDK node serialized by `std` should never be deserialized by `no-std` and vice versa. Seems like some weird negative durations could occur here in those situations

Reply:
Updated docs about mixing types being undefined. Also, moved `duration_since_epoch` into `Time` trait to make the association clearer.